modules/messages/server/services/message-to-influx.server.service.js

```diff
@@ -0,0 +1,186 @@
+
+// eslint complained here about unused variable => requiring without assignment
+require(path.resolve('./modules/messages/server/models/message.server.model'));
+
+
+var Message = mongoose.model('Message');
+
+
+/**
+ * This is a callback for the asynchronous influx modules
+ *
+ * @callback influxCallback
+ * @param {error} error
+ * @param {string} response
+ */
+
+/**
+ * This is a callback for the exports.process
+ *
+ * @callback processMessageCallback
+ * @param {error} error
+ * @param {object} fields - object of field keys and values as they'll be saved
+ * in influx
+ * @param {object} tags - object of tag keys and values as they'll be saved in
+ * influx
+ */
+
+/**
+ * this function is a shortcut for processing and sending the data to
+ * influxService (which sends it to influxdb)
+ * @param {object} message - a message object (as returned by mongoDB)
+ * @param {influxCallback} callback - a callback that handles the response
+ */
+module.exports.save = function (message, callback) {
+  async.waterfall([
+    function (done) {
+      return done(null, message);
+    },
+    module.exports.process,
+    module.exports.send
+  ], function (err, response) {
+    if (err) console.error(err);
+    if (typeof callback === 'function') {
+      return callback(err, response);
+    }
+  });
+};
+
+/**
+ * this function gets some info about the message sent and then calls the callback
+ * with field and tag keys and values as they're expected in database
+ * @param {object} message - a message object as returned by mongoDB
+ * @param {processMessageCallback} callback
+ */
+module.exports.process = function (message, callback) {
+  // some variables used later, filled inside waterfall
+  var isFirstMessage,
+      isFirstReply;
+
+  // fixing some strange filling of message data
+  // (userFrom is not id but user object)
+  var userFrom = message.userFrom._id
+    ? message.userFrom._id
+    : message.userFrom;
+  var userTo = message.userTo._id
+    ? message.userTo._id
+    : message.userTo;
+
+  async.waterfall([
+    function readFirstMessage(done) {
+      // find the oldest message of the thread
+      return Message.findOne({
+        $or: [
+          {
+            userTo: userTo,
+            userFrom: userFrom
+          },
+          {
+            userTo: userFrom,
+            userFrom: userTo
+          }
+        ]
+      })
+      .sort({ created: 1 })
+      .exec(done);
+    },
+
+    function readFirstReply(firstMessage, done) {
+      // if no message was found, throw error (there is always the first message
+      // already (at least the one just saved))
+      if (!firstMessage) {
+        var err = new Error('first message not found, but should have been already saved');
+        done(err);
+      }
+
+      // is the new message the first message of the thread?
+      isFirstMessage = !!(String(firstMessage._id) === String(message._id));
```

Comment:
How come we're casting (`!!`) a comparison here? The `!!` should have no effect as the result of `===` is always `true` or `false`.